File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/82930543256c5fba5011aa12d2578f54.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type DSum struct{}

func (ds *DSum) solve(in *fmt.Scanner, out *fmt.Writer) {
	var n, k int
	in.Scan(&n, &k)

	sums := make([]int64, n)
	arrs := make([][]int64, n)

	for i := 0; i < n; i++ {
		var t int
		in.Scan(&t)
		arrs[i] = make([]int64, t)
		for j := 0; j < t; j++ {
			in.Scan(&arrs[i][j])
		}
		if t > k {
			arrs[i] = arrs[i][:k]
		}
		sums[i] = sum(arrs[i])
	}

	dps := make([][]int64, 13)
	for i := range dps {
		dps[i] = make([]int64, k+1)
	}

	out.Println(rec(0, n, 0, k, arrs, sums, dps))
}

func sum(a []int64) int64 {
	var total int64
	for _, v := range a {
		total += v
	}
	return total
}

func doCopy(dpi int, dps [][]int64, k int) {
	copy(dps[dpi+1], dps[dpi])
}

func addToKnap(i, dpi int, k int, arrs [][]int64, sums []int64, dps [][]int64) {
	v := sums[i]
	k1 := len(arrs[i])
	for j := k; j >= k1; j-- {
		dps[dpi][j] = max(dps[dpi][j], dps[dpi][j-k1]+v)
	}
}

func rec(l, r, dpi, k int, arrs [][]int64, sums []int64, dps [][]int64) int64 {
	var ans int64
	carr := arrs[l]
	dp := dps[dpi]

	if l+1 == r {
		ans = max(ans, dp[k])
		var sum int64
		for i := 0; i < int(math.Min(float64(k), float64(len(carr)))); i++ {
			sum += carr[i]
			ans = max(ans, dp[k-i-1]+sum)
		}
	} else {
		m := (l + r) / 2
		doCopy(dpi, dps, k)
		for i := l; i < m; i++ {
			addToKnap(i, dpi, k, arrs, sums, dps)
		}
		ans = max(ans, rec(m, r, dpi+1, k, arrs, sums, dps))
		doCopy(dpi, dps, k)
		for i := m; i < r; i++ {
			addToKnap(i, dpi, k, arrs, sums, dps)
		}
		ans = max(ans, rec(l, m, dpi+1, k, arrs, sums, dps))
	}
	return ans
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func main() {
	var n, k int
	fmt.Scan(&n, &k)

	ds := DSum{}
	ds.solve(fmt.Scan, fmt.Println)
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/82930543256c5fba5011aa12d2578f54.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define MAX_N 100
#define MAX_K 10

typedef long long int ll;

ll max(ll a, ll b) {
    return (a > b) ? a : b;
}

ll sum(int arr[], int size) {
    ll sum = 0;
    for (int i = 0; i < size; i++) {
        sum += arr[i];
    }
    return sum;
}

ll rec(int l, int r, int k, int arrs[][MAX_K], ll sums[], ll dps[][MAX_K+1]) {
    ll ans = 0;
    int carr[MAX_K];
    ll dp[MAX_K+1];
    if (l+1 == r) {
        for (int i = 0; i <= k; i++) {
            dp[i] = dps[l][i];
        }
        smax(ans, dp[k]);
        ll sum = 0;
        for (int i = 0; i < min(k, (int)strlen(arrs[l])); i++) {
            carr[i] = arrs[l][i];
            sum += carr[i];
            smax(ans, dp[k-i-1] + sum);
        }
    } else {
        int m = (l+r)/2;
        for (int i = 0; i <= k; i++) {
            dp[i] = dps[l][i];
        }
        for (int i = l; i < m; i++) {
            ll v = sums[i];
            int k1 = strlen(arrs[i]);
            for (int j = k; j >= k1; j--) {
                smax(dp[j], dp[j-k1] + v);
            }
        }
        smax(ans, rec(m, r, k, arrs, sums, dp));
        for (int i = 0; i <= k; i++) {
            dp[i] = dps[l][i];
        }
        for (int i = m; i < r; i++) {
            ll v = sums[i];
            int k1 = strlen(arrs[i]);
            for (int j = k; j >= k1; j--) {
                smax(dp[j], dp[j-k1] + v);
            }
        }
        smax(ans, rec(l, m, k, arrs, sums, dp));
    }
    return ans;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);
    int arrs[MAX_N][MAX_K];
    ll sums[MAX_N];
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        for (int j = 0; j < t; j++) {
            scanf("%d", &arrs[i][j]);
        }
        if (t > k) {
            t = k;
        }
        sums[i] = sum(arrs[i], t);
    }
    ll dps[MAX_N][MAX_K+1];
    memset(dps, 0, sizeof(dps));
    printf("%lld\n", rec(0, n, k, arrs, sums, dps));
}

int main() {
    solve();
    return 0;
}
